src/main/scala/org/spartanz/parserz/ParsersModule2.scala

```diff
@@ -0,0 +1,326 @@
+      private[parserz] case class Sep[SI, SO, E, A, S](value: Grammar[SI, SO, E, A], sep: Grammar[SI, SO, E, S]) extends Grammar[SI, SO, E, SeparatedBy[A, S]]
+    }
+    // format: on
+
+    final val unit: Grammar[Any, Nothing, Nothing, scala.Unit] =
+      GADT.Produce(())
+
+    final def succeed[A](a: A): Grammar[Any, Nothing, Nothing, A] =
+      GADT.Produce(a)
+
+    final def fail[E, A](e: E): Grammar[Any, Nothing, E, A] =
+      unit.mapPartial(e)(PartialFunction.empty, PartialFunction.empty)
+
+    final def consume[E, A](e: E, codec: InOut.One[E, Input, A]): Grammar[Any, Nothing, E, A] =
+      GADT.Consume(codec.consumer, codec.condition, e)
+
+    final def consume[E, A](e: E, codec: InOut.Many[E, Input, A]): Grammar[Any, Nothing, E, A] =
+      GADT.ConsumeMany(codec.consumer, e)
+
+    final def consumeToken[E, A](e: E, codec: InOut.ManyConditional[E, Input, A]): Grammar[Any, Nothing, E, A] =
+      GADT.ConsumeToken(codec.consumer, e)
+
+    final def delay[SI, SO, E, A](g: => Grammar[SI, SO, E, A]): Grammar[SI, SO, E, A] =
+      GADT.Delay(() => g)
+
+    private def asEither[E, A, B](e: E)(f: A => Option[B]): A => E \/ B =
+      f(_).map(Right(_)).getOrElse(Left(e))
+  }
+
+  trait GrammarSyntax {
+
+    implicit final class ToStringOps1(self: String) {
+
+      def @@ [SI, SO, E, A](g: Grammar[SI, SO, E, A]): Grammar[SI, SO, E, A] =
+        g @@ self
+    }
+
+    implicit final class ToZipOps1[SI, SO, E, A, B](self: (Grammar[SI, SO, E, A], A)) {
+
+      def ~> [SI1 <: SI, SO1 >: SO, E1 >: E](that: Grammar[SI1, SO1, E1, B]): Grammar[SI1, SO1, E1, B] =
+        self._1.zipR(self._2, that)
+    }
+
+    implicit final class ToGrammarOps1[SI, SO, E, A, B](self: Grammar[SI, SO, E, List[A]]) {
+
+      def orEmpty: Grammar[SI, SO, E, List[A]] =
+        self.recover(Nil)
+    }
+
+    implicit final class ToGrammarOps2[SI, SO, E, A: ClassTag, B](self: Grammar[SI, SO, E, Array[A]]) {
+
+      def orEmpty: Grammar[SI, SO, E, Array[A]] =
+        self.recover(Array.empty[A])
+    }
+
+    implicit final class ToGrammarOps3[SI, SO, E, A, B](self: Grammar[SI, SO, E, (A, List[B])]) {
+
+      def foldLeft(fold: (A, B) => A, unfold: A =?> (A, B)): Grammar[SI, SO, E, A] =
+        self.map(
+          arg => {
+            arg._2.foldLeft(arg._1)(fold)
+          },
+          arg => {
+            @tailrec
+            def rec(acc: List[B])(a: A): (A, List[B]) =
+              unfold.lift(a) match {
+                case None          => (a, acc)
+                case Some((a1, b)) => rec(b :: acc)(a1)
+              }
+            rec(Nil)(arg)
+          }
+        )
+    }
+  }
+
+
+
+  private class ParserState[+E](val input: Input, var i: Int, var e: List[InOut.Error])
+
+  final def parser[S, E, A](grammar: Grammar[S, S, E, A]): Input => E \/ A =
+    input => step(grammar, new ParserState(input, 0, Nil))
+
+  private def step[S, E, A](grammar: Grammar[S, S, E, A], ps: ParserState[E]): E \/ A =
```

Comment:
Step should throw `E` values inside an exception type that does NOT store stack traces, to eliminate the overhead of `Either` boxing on each step.

Reply:
@jdegoes 
I did some experiments with this function.
1. Throwing exception instead of returning `Either` (basically as control flow) is about 2.5 times slower.
2. Returning `null` or `A` instead of `Either` is about 5% faster - hard to measure accurately.

The biggest perf improvement I had with this function so far is by avoiding lambdas and curried function signatures, specifically having a function with one parameters list returning a value (as opposed to several parameter lists and returning a function).

Reply:
What type of exception did you throw? With a stack trace or without?

Here's what I would suggest: rewrite `step` to be a `while` loop. Then you don't need `Either` or exceptions. You can use a product encoding of the either: a nullable variable that could be an error; and another nullable variable that could be the success.

Reply:
@jdegoes 
Exception without stack trace
```scala
private object ParserError extends Throwable with NoStackTrace
```
I already tried returning `null` or value instead of using `Either` or exception, it is only about 5% difference, not  significant.

I don't think while loop is possible since I need to analyze values produced by parts of `alt` and `zip` 🤔
But even if it is, what is the advantage?  I am basically reimplementing Java call stack, but all the code is the same, based on megamorphic function - the feature removed in FastParse 2. BTW, my current design is similar to FastParse 1 and sacrificing it made ~4x improvements possible in FastParse 2 😉

Reply:
I have uploaded the latest code.

Reply:
We could try to eliminate the megamorphic functions. To do that we have to eliminate _all_ lambdas and runtime functions such as those appearing in `Codec`. 

Essentially we'd be building a language for constructing data (case class / sealed trait). The language would including filtering, comparisons, basic math, etc. Then we can compile the parser to anything, including bytecode on the JVM.

How fast do you want to go? 😄